src/otaclient/client_package.py

```diff
@@ -0,0 +1,297 @@
+    PACKAGE_TYPE_PATCH = "patch"
+
+    def __init__(
+        self,
+        *,
+        base_url: str,
+        ota_metadata: OTAMetadata,
+        session_dir: StrOrPath,
+    ) -> None:
+        self._base_url = base_url
+        self._ota_metadata = ota_metadata
+        self._session_dir = Path(session_dir)
+        self._download_dir = self._session_dir / f".download_{os.urandom(4).hex()}"
+        self._download_dir.mkdir(exist_ok=True, parents=True)
+
+        self._manifest = None
+        self.package = None
+
+    @property
+    def _rootfs_url(self) -> str:
+        _metadata_jwt = self._ota_metadata.metadata_jwt
+        if _metadata_jwt is None or _metadata_jwt.rootfs_directory is None:
+            raise ValueError("metadata_jwt is not loaded yet, abort")
+        return urljoin_ensure_base(
+            self._base_url,
+            f"{_metadata_jwt.rootfs_directory.strip('/')}/",
+        )
+
+    def _prepare_manifest(
+        self,
+        condition: threading.Condition,
+    ) -> Generator[list[DownloadInfo]]:
+        """Download raw manifest.json and parse it."""
+        # ------ step 1: download manifest.json ------ #
+        _client_manifest_fpath = self._download_dir / Path(self.ENTRY_POINT).name
+        with condition:
+            yield [
+                DownloadInfo(
+                    url=urljoin_ensure_base(
+                        self._rootfs_url, self.ENTRY_POINT.lstrip("/")
+                    ),
+                    dst=_client_manifest_fpath,
+                )
+            ]
+            condition.wait()  # wait for download finished
+
+        # ------ step 2: load manifest.json ------ #
+        with open(_client_manifest_fpath, "r") as f:
+            manifest_data = json.load(f)
+            self._manifest = Manifest(**manifest_data)
+
+    def _prepare_client_package(
+        self,
+        condition: threading.Condition,
+    ) -> Generator[list[DownloadInfo]]:
+        """Download raw manifest.json and parse it."""
+
+        # ------ step 1: decide the target package ------ #
+        _available_package_metadata = self._get_available_package_metadata()
+
+        # ------ step 2: download the target package ------ #
+        _package_filename = _available_package_metadata.filename
+        _package_file = cfg.OTACLIENT_INSTALLATION_RELEASE + "/" + _package_filename
+        _downloaded_package_path = self._download_dir / _package_filename
+        with condition:
+            yield [
+                DownloadInfo(
+                    url=urljoin_ensure_base(
+                        self._rootfs_url, _package_file.lstrip("/")
+                    ),
+                    dst=_downloaded_package_path,
+                )
+            ]
+            condition.wait()  # wait for download finished
+        self.package = _available_package_metadata
+        self.downloaded_package_path = _downloaded_package_path
+
+    def _get_available_package_metadata(self) -> ReleasePackage:
```

Comment:
This is the part to decide the target package/patch from available list in manifest.